runtime/compiler/optimizer/J9EstimateCodeSize.cpp

```diff
@@ -411,6 +415,132 @@ TR_J9EstimateCodeSize::adjustEstimateForMethodInvoke(TR_ResolvedMethod* method,
+      {
+      int32_t argIndex = 0;
+      for (TR_ParameterMapping* parm = argMap.getFirst(); parm; parm = parm->getNext())
+         {
+         int32_t interimWeight = value;
+         TR::Node * parmNode = parm->_parameterNode;
+         TR::ParameterSymbol * parmSym = parm->_parmSymbol;
+         const char * parmClassName = NULL;
+         int32_t parmClassNameLen = 0;
+         char * argClassName = NULL;
+         TR_PrexArgument * prexArg = NULL;
+         if (prexArgInfo) prexArg = prexArgInfo->get(argIndex++);
+
+         if (parmSym)
+            parmClassName = parmSym->getTypeSignature(parmClassNameLen);
+
+         if (prexArg && prexArg->getClass())
+            argClassName = TR::Compiler->cls.classSignature(comp(), prexArg->getClass(), comp()->trMemory());
+
+         if (parmNode->getSymbolReference()
+            && parmNode->getSymbolReference()->getSymbol()->isStatic()
+            && parmNode->getSymbolReference()->getSymbol()->castToStaticSymbol()->isConstString())
+            {
+            value *= factor;
+            heuristicTrace(tracer(),"Setting size from %d to %d because arg is constant string.", interimWeight, value);
+            }
+         else if (argClassName
+                 && parmClassName
+                 && strncmp(argClassName, parmClassName, parmClassNameLen) != 0 // arg type needs to be more specific than declared type to benefit from weight adjustment
+                 && strncmp(parmClassName, "Ljava/lang/String;", parmClassNameLen) == 0)
+            {
+            value *= lessAggressiveAdjustmentFactor;
+            heuristicTrace(tracer(),"Setting size from %d to %d because arg is a string.", interimWeight, value);
+            }
+         else if (parmNode->getOpCodeValue() == TR::aload
+                 && parmNode->getSymbolReference()
+                 && parmNode->getSymbolReference()->getSymbol()->isConstObjectRef())
+            {
+            value *= factor;
+            heuristicTrace(tracer(),"Setting size from %d to %d because arg is const object ref.", interimWeight, value);
+            }
+         else if (parmNode->getOpCodeValue() == TR::aloadi
+               && parmNode->getSymbolReference()
+               && parmNode->getSymbolReference() == comp()->getSymRefTab()->findJavaLangClassFromClassSymbolRef())
+            {
+            if (parmNode->getFirstChild()
+               && parmNode->getFirstChild()->getOpCodeValue() == TR::loadaddr
+               && parmNode->getFirstChild()->getSymbol()->isStatic()
+               && !parmNode->getFirstChild()->getSymbolReference()->isUnresolved()
+               && parmNode->getFirstChild()->getSymbol()->isClassObject())
+               {
+               value *= factor;
+               heuristicTrace(tracer(),"Setting size from %d to %d because arg is const class ref.", interimWeight, value);
+               }
+            else
+               {
+               value *= lessAggressiveAdjustmentFactor;
```

Comment:
This path still applies the less-aggressive adjustment factor (based on our knowledge of the type of `<javaLangClassFromClass>`) even if the parameter type is `java/lang/Class`

Reply:
Removed this path.

Reply:
I thought you'd check `parmClassName`, but I agree it makes sense that `prexArg` is where we should have the knowledge that an argument coming from `<javaLangClassFromClass>` is a `java/lang/Class` (even if `prexArg` doesn't manage to capture that at the moment)